File: FirstLastPosition.cpp
// Time Complexity : O(2logN)
// Space Complexity : O(1) as I didn't use any extra space, just the temporary variables
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Nope
// Problem Link: https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
// Submission Link: https://leetcode.com/submissions/detail/527292311/

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> indices(2, -1);
        int start = 0, end = nums.size() - 1;
        int mid;
        
        while(start <= end){
            mid = (start + end) / 2;
            
            if(nums[mid] == target){
                indices[0] = mid;
                end = mid - 1;
            }
            else if(target < nums[mid])
                end = mid - 1;
            else
                start = mid + 1;
        }
        
        start = 0, end = nums.size() - 1;
        
        while(start <= end){
            mid = (start + end) / 2;
            
            if(nums[mid] == target){
                indices[1] = mid;
                start = mid + 1;
            }
            else if(target < nums[mid])
                end = mid - 1;
            else
                start = mid + 1;
        }
        
        return indices;
    }
};

File: FirstLastPosition.cs
// Time complexity: O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode: Yes (https://leetcode.com/submissions/detail/569648307/)
// Any problem you faced while coding this:
namespace BinarySearch2
{
    class FirstLastPosition 
    {        
        public int[] SearchRange(int[] nums, int target) {
            int left = 0;
            int right = nums.Length - 1;
            int[] result = new int[2]{-1, -1};
            int foundFlag = 0;

            if(nums.Length < 1 || target < nums[left] || target > nums[right]) {
                return result;
            }

            //Finding left most element
            while(left <= right) {
                int mid = left + (right - left)/2;

                if(nums[mid] < target) {
                    left = mid + 1;
                } else if (nums[mid] > target) {
                    right = mid - 1;
                } else {
                    right = mid - 1;
                    foundFlag = 1;
                }
            }
            if(foundFlag == 1) {
                result[0] = left;
            } else {
                return result;
            }

            left = 0;
            right = nums.Length - 1;

            //Finding right most element
            while(left <= right) {
                int mid = left + (right - left)/2;

                if(nums[mid] > target) {
                    right = mid - 1;
                } else if (nums[mid] < target) {
                    left = mid + 1;
                } else {
                    left = mid + 1;
                }
            }

            result[1] = right; 
            return result;
        }

        static void Main(string[] args) {
            FirstLastPosition find = new FirstLastPosition();
            int[] nums = new int[]{5, 7, 7, 8, 8, 10};
            int[] result = find.SearchRange(nums, 8);
            Console.Write(result[0] + ",");
            Console.Write(result[1]);
        }
    }
}
